Controller & Actions for Pro request process

stevenday opened the pull request:
So far I have:
- A draft request model
- A new request form
- A create action
- A "preview" page
- An edit form
- An update action

Next is the big question, what happens when we "send" the request.

You can see that I've tried to simplify the data model a bit by moving the
message body into the Draft, rather than creating an OutgoingMessage alongside
it. My working assumption was that next I'd add a method to InfoRequest called
`new_from_draft` or something similar. This could do the transfer of data,
make an outgoing message an so on. Alternatively, this could happen in a new
controller (we'll need one anyway, so it's maybe a moot point).

My main question is what happens with any validation errors. Right now both
the message and the request do some, so we'd need a way to combine that and
feed it back to the draft I guess?

After that, there are a few outstanding issues with stuff I took from the
current new request form, like how we display the `law_used_short` text etc.

stevenday commented:
I've now added a really basic version of the publishing process to do what I've described above. It's in the controller for now, as it cuts across three models so it seems odd to put it any one of them and worse to spread it across all three. I haven't extensively tested the validation/errors side of things and that's the area I think will need some more thought. Any ideas how to make that generic/future-proof? I guess there's something meta I can do looping over all the errors and checking which ones match a property of the draft object?

garethrees commented on app/controllers/alaveteli_pro/draft_info_requests_controller.rb:
```diff
@@ -0,0 +1,121 @@
+
+  def update
+    @draft_info_request = DraftInfoRequest.find(params[:id])
+    @draft_info_request.update_attributes(draft_info_request_params)
```
update_attributes calls save.

You want:

```ruby
if @draft_info_request.update_attributes(draft_info_request_params)
  # ...
```

garethrees commented on app/controllers/alaveteli_pro/draft_info_requests_controller.rb:
```diff
@@ -0,0 +1,121 @@
+
+  def create
+    @draft_info_request = DraftInfoRequest.new(draft_info_request_params)
+    @draft_info_request.user = current_user
```
Instead of this you can (and should) do:

```ruby
@draft_info_request = current_user.draft_info_requests.new(draft_info_request_params)
```

garethrees commented on app/controllers/alaveteli_pro/draft_info_requests_controller.rb:
```diff
@@ -0,0 +1,121 @@
+    @draft_info_request = DraftInfoRequest.find(params[:id])
+  end
+
+  def publish
```
I think we should try to stick to the default RESTful actions for new controllers.

I'd make this something like `DraftPublicationsController#create`.

garethrees commented on app/controllers/alaveteli_pro/draft_info_requests_controller.rb:
```diff
@@ -0,0 +1,121 @@
+
+  def publish
+    @draft_info_request = DraftInfoRequest.find(params[:id])
+    build_info_request_and_outgoing_message(@draft_info_request)
```
I'd also add `InfoRequest.from_draft` that initializes a new `InfoRequest` from the draft:
 rather than keeping this logic in the controller:

```ruby
class InfoRequest
  def self.from_draft(draft)
    request = new(:title => draft.title,
                  :user => draft.user,
                  :public_body => draft.public_body)
    request.outgoing_messages.new(:body => draft.body,
                                  :status => 'ready',
                                  :message_type => 'initial_request',
                                  :what_doing => 'normal_sort')
    request
  end
```

garethrees commented on app/controllers/alaveteli_pro/draft_info_requests_controller.rb:
```diff
@@ -0,0 +1,121 @@
+# -*- encoding : utf-8 -*-
+# app/controllers/alaveteli_pro/request_controller.rb
+# Dashboard controller, for pro user dashboards.
```
copy/paste error?

garethrees commented on app/controllers/alaveteli_pro/draft_info_requests_controller.rb:
```diff
@@ -0,0 +1,121 @@
+  end
+
+  def edit
+    @draft_info_request = DraftInfoRequest.find(params[:id])
```
Should scope to `current_user`, otherwise you can just bash in ID numbers to the URL and load other users' drafts:

```ruby
@draft_info_request = current_user.draft_info_requests.find(params[:id])
```

Should do this for all actions

stevenday commented:
@garethrees, @crowbot thanks for the comments! A few thoughts in return after I've worked through your suggestions:

> We could put the validations that apply to both draft requests and requests in a concern or custom validator that applies to both

> I'm wondering whether a Draft should also initialize an internal InfoRequest so that methods like law_used can be passed over to that.

In either of these cases it's complicated a bit because I've merged an OutgoingMessage and an InfoRequest into one model, so we need some of the validation from both of them. I tried extracting the validation into two concerns, but OutgoingMessage's validation is pretty complicated because of the use of the templates, which makes it quite hard to extract. The alternative of keeping an internal model for each seems more do-able though, I'll flesh that out. (This also gets us easier access to the law_used stuff, so that's good)

> I think we should try to stick to the default RESTful actions for new controllers.
> I'd make this something like DraftPublicationsController#create.

> In terms of the conversion of a draft to a request, it feels to me like this is, as discussed above, really the creation of an InfoRequest using a DraftInfoRequest. How about we use a create action on the RequestController to do this, passing the id of the draft?

I think I prefer the latter suggestion here, as `DraftPublicationsController#create` seems a bit contrived? Keeping it inside the pro namespace seems like the best idea for now, so that we don't mix up the draft request functionality until it's meant to be available for the public too.

garethrees commented:
I think a [sketch of the UI flow](https://signalvnoise.com/posts/1926-a-shorthand-for-designing-ui-flows) would help figure out how we want to do this.

Maybe we don't ever want to go straight from a saved draft to a published InfoRequest. I'm thinking now that editing a draft would actually just be hitting `RequestController#new` with some pre-filled data like http://alaveteli.org/docs/developers/api/#starting-new-requests-programmatically.

If we did this, then we could eliminate `edit`, `show`, `publish` and possibly `update` from the `DraftInfoRequestsController`.

stevenday commented:
I've just pushed a load of commits with some updates to this. I've implemented the `InfoRequest#from_draft` method and @garethrees internal request/message object, with some method shadowing to make the titles work and make sure that the draft keeps it's internal objects up to date. I've also pushed the InfoRequest creation to a new controller. I took some ideas from @crowbot validation, making it possible to choose whether the draft validates its internal objects or not. There's still the downside of copying across errors, but it seems like we're stuck in a similar situation regardless because otherwise we're copying across (or including) *some* validation but not all.

Re: the UI flow, here's a gif of how it works currently:
![](http://g.recordit.co/ZWEmI0w0K2.gif)

As you can see, I've added a new "Save Draft" button, with the intention being that we'd also expand this into some auto-saving JS eventually. I've slightly renamed the Preview button to make it clearer that it's kind of a "next step" so the validation is less of a shock, but I wonder whether this step is even necessary for Pros? You don't "preview" an email, so should you preview this?

If this is generally good, then I'll start writing some specs for all the new things and tidy up the commits.

garethrees commented:
> Re: the UI flow, here's a gif of how it works currently:

Looking solely at the gif, its not clear why we need any of the `DraftInfoRequestsController` actions other than `#create`.

Why do we need a separate form etc? Seems to be a lot of duplication (and complexity in `DraftInfoRequest`) that we might be able to avoid.

stevenday commented:
> Why do we need a separate form etc? Seems to be a lot of duplication (and complexity in DraftInfoRequest) that we might be able to avoid.

So this is just the basic version where it looks a lot like the current request form, but soon we'll be adding embargo options (which need to do some funky stuff to set account/project defaults etc), project filing and the js autocomplete authority picker, which will differentiate it quite a bit from the current request flow.

The idea of a draft request is twofold: firstly so that we can actually let people build up requests without having to send them immediately and secondly so that we can deal with the problem of a non-pro writing a request anonymously and then logging in - because we need to send them somewhere to set the embargo, project etc. 

> Looking solely at the gif, its not clear why we need any of the DraftInfoRequestsController actions other than #create and #index.

I'm not sure I follow here, I think this way of doing it (with separate controller actions for new, edit, create, update etc) is clearer and more idiomatic.. 

I admit that the model is a bit tricky now, I'd be happy to have another go at extracting the validation stuff from existing models if that seemed preferable?

stevenday commented:
Here's a UI flow diagram as I see it at the mo:

![img_0139](https://cloud.githubusercontent.com/assets/663700/20532974/cf6cece2-b0d3-11e6-86be-de1e64ac2dee.jpg)

garethrees commented:
Here's what I was thinking we could do:

![draft-creation-flow](https://cloud.githubusercontent.com/assets/282788/20533322/fe89050a-b0d4-11e6-9be3-fd8b9f7d2036.jpg)

The top line is the current "make a request process" as it is now.

The new addition would be an option to click "save as draft" in step 2. This would post the parameters to `InfoRequestDraftsController#create` and save the title, body and public body ID.

When the user wants to continue with their draft, they'd visit `InfoRequestDraftsController#index` to find the draft and click "Edit draft". This would actually be a link to `RequestController#new` with the ["API" parameters](http://alaveteli.org/docs/developers/api/#starting-new-requests-programmatically) pre-set from the persisted draft.

They could then continue through the process as normal.

stevenday commented:
@garethrees - aha, gotcha. 

Hmm, the downside of this for me is that it forces you to make a valid request (title, body and public_body) before you can save it as a draft. I like the idea of being able to save something straight away, even if it's just a title or a selected public body.

As I said above, we're probably going to need a new `InfoRequestController#create` and `#new` for the new pro options anyway, so does this make as much sense then?

garethrees commented:
> Hmm, the downside of this for me is that it forces you to make a valid request (title, body and public_body) before you can save it as a draft. I like the idea of being able to save something straight away, even if it's just a title or a selected public body.

Not necessarily – when you POST to `DraftsController#create` you'd still be creating an `InfoRequestDraft` – which could allow all attributes to be blank. It won't matter if they're blank when you use the API params, as they just won't pre-fill the blank fields.

> As I said above, we're probably going to need a new `InfoRequestController#create` and `#new` for the new pro options anyway, so does this make as much sense then?

Yeah, this comment…

> So this is just the basic version where it looks a lot like the current request form, but soon we'll be adding embargo options (which need to do some funky stuff to set account/project defaults etc), project filing and the js autocomplete authority picker, which will differentiate it quite a bit from the current request flow.

…made me think that actually, what this PR is discussing is almost that – a new request process for pro.

In which case, yes, I'd be thinking about making `AlaveteliPro::RequestController` for whatever actions need it, and keep the drafting pretty simple as I illustrated.

I think the only complication is if you wanted to store the embargo with the draft, but tbh I'm not sure how valuable that is. I think you're more likely to want to set it at the point of actually creating the request?

garethrees commented:
> I think the only complication is

Oh, you'd require a public body, but it looks like you were validating that existed already.

stevenday commented:
> Not necessarily – when you POST to DraftsController#create you'd still be creating an InfoRequestDraft – which could allow all attributes to be blank. It won't matter if they're blank when you use the API params, as they just won't pre-fill the blank fields.

OK, currently clicking the preview button does the validation though, rather than validating after you click send, so we'd have to change that. I think that would be a (minor) step backwards in UX terms.

> Oh, you'd require a public body, but it looks like you were validating that existed already.

It isn't required on the draft model so it shouldn't be validated until you try to send it, I just dropped a `<select>` in the html to pave the way for the autocomplete version and didn't give it a blank default `<option>`.


I'm still not totally sold on the value of re-using the new form to edit drafts. 

I don't think we're going to want to use the existing form anyway because of the amount of new stuff we're adding (only the title and body fields will remain the same), so what difference does it make whether we're creating/editing a draft and then POSTing to create an InfoRequest, or vice-versa?

(I'm not sure how re-editing a draft would work in your proposal, but presumably it would need an #update or a more complicated #create to deal with it? Either way it doesn't seem to get us much less or simpler code than what we have currently.)

garethrees commented on app/models/draft_info_request.rb:
```diff
@@ -0,0 +1,80 @@
+    if @info_request.nil?
+      @info_request = InfoRequest.from_draft(self)
+    end
+    @info_request
```
You can make this much simpler:

```ruby
  def info_request
    @info_request ||= InfoRequest.from_draft(self)
  end
```

garethrees commented:
> OK, currently clicking the preview button does the validation though, rather than validating after you click send, so we'd have to change that. I think that would be a (minor) step backwards in UX terms.

I was imagining you save the draft before going through to the preview, but ISWYM.

I was mostly thinking of this workflow being like how Sparrow works:

![sparrow-draft-ui](https://cloud.githubusercontent.com/assets/282788/20535252/013f2e94-b0dc-11e6-83ce-a4f87183c6b6.gif)

> (I'm not sure how re-editing a draft would work in your proposal, but presumably it would need an #update or a more complicated #create to deal with it? Either way it doesn't seem to get us much less or simpler code than what we have currently.)

Yeah, that might be a bit trickier. With this workflow I'd maybe make the "edit" button hit `GET /new/:body_url_name?draft_id=4`, and then set the API params from the draft record. That way you could use the ID when you update the draft, or delete the draft once you've clicked "send". But yeah, a little more complex.

> I don't think we're going to want to use the existing form anyway because of the amount of new stuff we're adding (only the title and body fields will remain the same), so what difference does it make whether we're creating/editing a draft and then POSTing to create an InfoRequest, or vice-versa?

Here are my main thoughts, but I'm not dead against what you've got here.

- What I've suggested could be applied to Alaveteli core with minimum effort, and isn't tied to the pro-only request process
- The less actions in `DraftsController`, the less a user needs to think about. Why should editing a "draft" be different to editing a "real" request.
- The composition of the current `InfoRequestDraft` is quite complicated. I'm not 100% convinced a draft needs to know much more than title, body and public body – all of which are pretty optional.

stevenday commented:
ok, here goes again :)

I've moved the action into an `InfoRequestController` as suggested by @crowbot. This has three actions, `#new` for the creating/editing form, `#preview` for previewing the request and `#create` for actually making it. This has allowed me to remove all the fancy validation stuff from `DraftInfoRequest`, and actually get rid of the `DraftInfoRequestController` altogether.

The icky part is that `InfoRequestController#preview` has to know about creating/updating drafts, because the `#new` form can only post to one place. I don't think it's turned out too horribly, but I'd appreciate it if you can point out anything obvious I've missed there which would make it nicer? One upside is that I think it makes previewing stuff much simpler, because we can save the draft and then just recreate the form from that. As a side-effect, if you forget about your request whilst previewing, it'll still be there as a draft. Again, once the `#new` form auto-saves too (or saves via an ajax post at least) this will be a bit clearer.

Perhaps I should formalise the relationship between draft and info_request with a `has_one` relationship? (Possibly doing away with destroying the draft after sending too?)

stevenday commented:
I've also just pushed the embargo side of this, it's a pretty simple addition in the same vein as the way outgoing messages are treated, with a field to cache the important stuff on the draft, and then a little bit of controller work to create/validate an embargo object in the various actions and views.

stevenday commented:
> Ah! Perhaps we can have the best of both worlds - the 'new' form can be submitted to the draft controller 'create' or 'update' action (depending on whether there's a draft id in the form), but if the params include 'preview' ie. the preview button was used to submit the form, then those actions redirects to the 'preview' action of the InfoRequestController, not the 'new' action. The preview controller creates an InfoRequest etc from the draft and validates them, then continues as it does now.

I've just updated the code to match this description as best I can, and re-worked the embargo setting to use a text duration that only gets turned into a `publish_at` date on the Embargo when it's saved. I've also rebased everything and tidied it up as if this was the plan all along.

I'm leaving embargo extending/altering and the authority autocomplete until later, as this is getting quite big already and I don't want to drag it on any longer than necessary.

@crowbot - this is hopefully ready for a final review now

stevenday commented on app/models/user.rb:
```diff
@@ -32,6 +33,7 @@
 #  request_classifications_count     :integer          default(0), not null
 #  public_body_change_requests_count :integer          default(0), not null
 #  info_request_batches_count        :integer          default(0), not null
+#  pro                               :boolean          default(FALSE), not null
```
Oops, thanks for spotting that, I've dropped the annotation commit

stevenday commented on spec/models/draft_info_request_spec.rb:
```diff
@@ -0,0 +1,26 @@
+
+describe DraftInfoRequest do
+  it "belongs to a public body" do
+    expect(FactoryGirl.create(:draft_info_request).public_body).not_to be nil
```
Done in 0e1880e

stevenday commented on app/models/embargo.rb:
```diff
@@ -1,16 +1,56 @@
+    "3_months" => Proc.new { 3.months },
+    "6_months" => Proc.new { 6.months },
+    "12_months" => Proc.new { 12.months },
+    "indefinitely" => Proc.new { 1000.years }
```
Done in 93b3eb4

stevenday replied:
and c3f2823

stevenday commented on app/controllers/alaveteli_pro/info_requests_controller.rb:
```diff
@@ -0,0 +1,96 @@
+    (@embargo.nil? || @embargo.present? && @embargo.valid?)
+  end
+
+  def set_publish_at_options
```
Done in 1d8e80d

stevenday commented on spec/integration/alaveteli_pro/request_creation_spec.rb:
```diff
@@ -0,0 +1,209 @@
+          click_button "Save draft"
+
+          # Redirected back to new request form
+          draft = DraftInfoRequest.where(title: "Does the pro request form work?")
```
done in f9ca9dc

stevenday commented on spec/integration/alaveteli_pro/request_creation_spec.rb:
```diff
@@ -0,0 +1,209 @@
+    it "allows us to preview the request" do
+      with_feature_enabled(:alaveteli_pro) do
+        using_session(pro_user_session) do
+          # New request form
```
Done in c5cf2d6
